### ISSUE TITLE
Indication listener port reuse fails in Python 3

### ISSUE THREAD

**Issue by KSchopmeyer** [pywbem/pywbem#452]
I reran the tests that are causing the problem several times.  I have concluded that this may be an issue with the test code in that I do not force the listener stop if there is an issue with any of the indication tests.

We sometimes get the following error back from the listener when it is being used in run_cimoperations.py

Traceback (most recent call last):
  File "testsuite/run_cim_operations.py", line 6301, in test_not_owned_indications2
    https_port=https_listener_port)
  File "testsuite/run_cim_operations.py", line 5797, in create_listener
    my_listener.start()
  File "/home/kschopmeyer/pywbem/githubpywbem/pywbem/pywbem/_listener.py", line 752, in start
    ListenerRequestHandler)
  File "/usr/lib/python3.4/socketserver.py", line 430, in __init__
    self.server_bind()
  File "/usr/lib/python3.4/http/server.py", line 133, in server_bind
    socketserver.TCPServer.server_bind(self)
  File "/usr/lib/python3.4/socketserver.py", line 444, in server_bind
    self.socket.bind(self.server_address)
OSError: [Errno 98] Address already in use

I have been unable to duplicate the issue in test_indication listener despite the fact that it starts and stops the listener multiple times. Not sure yet whether this is an issue with my test code or a real issue in the listener

I tried the following test: Using test_indication.py I ran it without calling listener.stop().  The next time I called the test, it responded with:

Traceback (most recent call last):
  File "testsuite/test_indicationlistener.py", line 243, in test_send_100
    self.send_indications(100, 50000, None)
  File "testsuite/test_indicationlistener.py", line 186, in send_indications
    self.createlistener(host, http_port)
  File "testsuite/test_indicationlistener.py", line 167, in createlistener
    LISTENER.start()
  File "/home/kschopmeyer/pywbem/githubpywbem/pywbem/pywbem/_listener.py", line 752, in start
    ListenerRequestHandler)
  File "/usr/lib/python3.4/socketserver.py", line 430, in __init__
    self.server_bind()
  File "/usr/lib/python3.4/http/server.py", line 133, in server_bind
    socketserver.TCPServer.server_bind(self)
  File "/usr/lib/python3.4/socketserver.py", line 444, in server_bind
    self.socket.bind(self.server_address)
OSError: [Errno 98] Address already in use

Since this does a test and then immediatly another test, it would appear that the port is kept open between the tests.   Thus the first test passes but the second test fails because the listener.stop() has not been executed.   However, when I run the test repeatedly, it does not cause a failure for the first test of each run but obviously does for the second test.  

Thus this shows the same result in test_indications.py.   There is something different between the pegasus tests and the test_indication tests.
  

Created a new test that starts and stops the listener for multiple tests. 

I have concluded that this is an issue between OpenPegasus and pywbem because of my test code.  In case of failure of an indication test I am not forcing the listener closed.  So the next test fails.

It works on python2 but gets the following error on python3

self = <pywbem._listener.ThreadedHTTPServer object at 0x2b105847de48>

```
def server_bind(self):
    """Called by constructor to bind the socket.

        May be overridden.

        """
    if self.allow_reuse_address:
        self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
```

> ```
>   self.socket.bind(self.server_address)
> ```
> 
> E       OSError: [Errno 98] Address already in use

/usr/lib/python3.4/socketserver.py:444: OSError

Which means that we should be doing more I guess when we stop the listener.

I set this one to 0.9.0 moptional but we will need to set up a warning for 0.9.0

**Comment by KSchopmeyer** [comment]
I think I may have a handle on the problem.  Clearly, when we shutdown, the http server it is not immediatly closing the tcp connection and therefore causing the issue:

shutdown the server, and do socket.close(), may leave some connections  TIME_WAIT state. This throws address already in use.

In theory, that is indeed the reason for the TIME_WAIT state.  Using SO_REUSEADDR is pretty safe, and common practice before the tcp bind.

**Comment by KSchopmeyer** [comment]
Karl will write test to confirm that this fixes issue.

**Comment by andy-maier** [comment]
The scenario in which it failed was:
* First `WBEMListener` object created that is listening on a particular port, it is started and stopped.
* New `WBEMListener` object created that is listening on the same port. When started, it fails.

**Comment by KSchopmeyer** [comment]
I propose that this issue be superceded by issue #1090.  While there is actually a hole in python that could cause this to happen, what I was seeing in run_cimoperations.py happening frequently was really the issue around always stopping the port.  I am cleaning up run_cimoperations.py to avoid this issue.

**Comment by KSchopmeyer** [comment]
We fixed the run_cimoperations.py issues with issue #1090 and I therefore propose that we can close this
issue as resolved by #1090

**Comment by andy-maier** [comment]
Closing this issue because it is not handled in #1090.
